--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1450,6 +_,8 @@
             return false;
         }
 
+        com.infernalsuite.asp.AdvancedSlimePaper.instance().onWorldUnload(world.getName()); // ASP - Remove unloaded world from map
+
         try {
             if (save) {
                 handle.save(null, true, false); // Paper - Fix saving in unloadWorld
