--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -350,6 +_,12 @@
                 if (flag2) {
                     lightEngine.queueSectionData(LightLayer.SKY, sectionPos, sectionData.skyLight);
                 }
+
+                //ASP start
+                if (level instanceof com.infernalsuite.asp.level.SlimeLevelInstance) {
+                    poiManager.checkConsistencyWithBlocks(sectionPos, sectionData.chunkSection);
+                }
+                //ASP end
             }
         }
 
