From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Date: Mon, 26 Dec 2022 11:25:35 -0500
Subject: [PATCH] World overrides


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 3349baa2ab971db435b88083688bd0f99474607e..610e58bede7e1482fc861ab6559f4f3a6e7e3633 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -516,18 +516,33 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             if (levelStemKey == LevelStem.NETHER) {
                 if (this.server.getAllowNether()) {
                     dimension = -1;
+                    // ASP START
+                    if(com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadNetherOverride()) {
+                        continue;
+                    }
+                    // ASP END
                 } else {
                     continue;
                 }
             } else if (levelStemKey == LevelStem.END) {
                 if (this.server.getAllowEnd()) {
                     dimension = 1;
+                    // ASP START
+                    if(com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadEndOverride()) {
+                        continue;
+                    }
+                    // ASP END
                 } else {
                     continue;
                 }
             } else if (levelStemKey != LevelStem.OVERWORLD) {
                 dimension = -999;
             }
+            // ASP START
+            else if(levelStemKey == LevelStem.OVERWORLD && com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadOverworldOverride()) {
+                continue;
+            }
+            // ASP END
 
             // Migration of old CB world folders...
             String worldType = (dimension == -999) ? levelStemKey.location().getNamespace() + "_" + levelStemKey.location().getPath() : org.bukkit.World.Environment.getEnvironment(dimension).toString().toLowerCase(Locale.ROOT);
