From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Mayr <davidliebtkekse@gmail.com>
Date: Thu, 13 Mar 2025 00:09:20 +0100
Subject: [PATCH] Read only dimension data store


diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index af67b07722bb0125acd081dab767d7e7b360623b..fc4f631f7d55ce67f348e3b6090095e6ef8445e8 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -208,7 +208,13 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             LOGGER.error("Failed to create dimension data storage directory", (Throwable)var14);
         }
 
-        this.dataStorage = new DimensionDataStorage(path, fixerUpper, level.registryAccess());
+        //ASP start - No dimension data storage
+        if(level instanceof com.infernalsuite.asp.level.SlimeLevelInstance) {
+            this.dataStorage = new com.infernalsuite.asp.level.ReadOnlyDimensionDataStorage(path, fixerUpper, level.registryAccess());
+        } else {
+            this.dataStorage = new DimensionDataStorage(path, fixerUpper, level.registryAccess());
+        }
+        //ASP end - No dimension data storage
         this.ticketStorage = this.dataStorage.computeIfAbsent(TicketStorage.TYPE);
         this.chunkMap = new ChunkMap(
             level,
