--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -313,6 +_,13 @@
         }
     }
 
+
+    // ASWM start - maintain binary compatibility with 1.20.6
+    public ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder getChunkHolder() {
+        return this.level.moonrise$getChunkTaskScheduler().chunkHolderManager.getChunkHolder(chunkPos.x, chunkPos.z);
+    }
+    // ASWM end
+
     // Paper start - If loaded util
     @Override
     public final FluidState getFluidIfLoaded(BlockPos blockposition) {
