--- a/ca/spottedleaf/moonrise/paper/PaperHooks.java
+++ b/ca/spottedleaf/moonrise/paper/PaperHooks.java
@@ -79,7 +_,10 @@
 
     @Override
     public void chunkUnloadFromWorld(final LevelChunk chunk) {
-
+        if (chunk.level instanceof com.infernalsuite.asp.level.SlimeLevelInstance slime) {
+            //The custom entity slices need to be passed on for entities to be saved
+            slime.onChunkUnloaded(chunk, chunk.getChunkHolder().getEntityChunk());
+        }
     }
 
     @Override
