From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Mayr <davidliebtkekse@gmail.com>
Date: Mon, 10 Mar 2025 12:09:41 +0100
Subject: [PATCH] Delete temp folder after world is unloaded


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 185bfb619767a2ce8b831b083d946e9f8e7bfb0c..a920263b6eef436abe113d1264a782e458748728 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1460,6 +1460,12 @@ public final class CraftServer implements Server {
             handle.getChunkSource().close(save);
             io.papermc.paper.FeatureHooks.closeEntityManager(handle, save); // SPIGOT-6722: close entityManager // Paper - chunk system
             handle.levelStorageAccess.close();
+
+            //ASP start - avoid temp storage leak during runtime
+            if(handle instanceof com.infernalsuite.asp.level.SlimeLevelInstance asp) {
+                asp.deleteTempFiles();
+            }
+            //ASP end
         } catch (Exception ex) {
             this.getLogger().log(Level.SEVERE, null, ex);
         }
