From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Mayr <davidliebtkekse@gmail.com>
Date: Mon, 10 Mar 2025 12:09:41 +0100
Subject: [PATCH] Delete temp folder after world is unloaded


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 4b4dc5ff38f74a008b5b0bebe312fbbf0ace7b7c..c0358d9291b8b569cd234f9a3e64f5ad60054377 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1519,6 +1519,12 @@ public final class CraftServer implements Server {
             handle.getChunkSource().close(save);
             io.papermc.paper.FeatureHooks.closeEntityManager(handle, save); // SPIGOT-6722: close entityManager // Paper - chunk system
             handle.levelStorageAccess.close();
+
+            //ASP start - avoid temp storage leak during runtime
+            if(handle instanceof com.infernalsuite.asp.level.SlimeLevelInstance asp) {
+                asp.deleteTempFiles();
+            }
+            //ASP end
         } catch (Exception ex) {
             this.getLogger().log(Level.SEVERE, null, ex);
         }
