From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Mayr <davidliebtkekse@gmail.com>
Date: Wed, 12 Mar 2025 21:14:56 +0100
Subject: [PATCH] Prevent config disk io on world load


diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index 025ec3f3e84a3367b4d709c8c3ecd98f6dfecab4..182de15ca133fc42bf703ee9d116302bf5207c0d 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -79,7 +79,12 @@ public class SpigotConfig {
         }
     }
 
+    // ASP start - Improve Slime IO
     public static void readConfig(Class<?> clazz, Object instance) {
+        readConfig(clazz, instance, true);
+    }
+    public static void readConfig(Class<?> clazz, Object instance, boolean save) {
+    // ASP end - Improve Slime IO
         for (Method method : clazz.getDeclaredMethods()) {
             if (Modifier.isPrivate(method.getModifiers())) {
                 if (method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE) {
@@ -96,7 +101,11 @@ public class SpigotConfig {
         }
 
         try {
-            SpigotConfig.config.save(SpigotConfig.CONFIG_FILE);
+            // ASP start - Improve Slime IO
+            if(save) {
+                SpigotConfig.config.save(SpigotConfig.CONFIG_FILE);
+            }
+            // ASP end - Improve Slime IO
         } catch (IOException ex) {
             Bukkit.getLogger().log(Level.SEVERE, "Could not save " + SpigotConfig.CONFIG_FILE, ex);
         }
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index 339832e312578ee385bbf3c9d5c131435abe7d7a..3924ba62c2df8372b67f6666f0c02d007ea3021c 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -15,13 +15,23 @@ public class SpigotWorldConfig {
     private boolean verbose;
 
     public SpigotWorldConfig(String legacyWorldName, Key worldKey) {
+        // ASP start - Improve Slime IO
+        this(legacyWorldName, worldKey, true);
+    }
+
+    public SpigotWorldConfig(String legacyWorldName, Key worldKey, boolean saveOnLoad) {
+        // ASP end - Improve Slime IO
         this.legacyWorldName = legacyWorldName;
         this.worldName = worldKey.asString();
         this.config = SpigotConfig.config;
-        this.init();
+        this.init(saveOnLoad);  // ASP - Improve Slime IO
     }
-
     public void init() {
+        // ASP start - Improve Slime IO
+        init(true);
+    }
+    public void init(boolean saveOnLoad) {
+        // ASP end - Improve Slime IO
         this.verbose = this.getBoolean("verbose", false); // Paper
         if (SpigotConfig.version <= 12) {
             ConfigurationSection section = this.config.getConfigurationSection("world-settings." + this.legacyWorldName);
@@ -33,7 +43,7 @@ public class SpigotWorldConfig {
         }
 
         this.log("-------- World Settings For [" + this.worldName + "] --------");
-        SpigotConfig.readConfig(SpigotWorldConfig.class, this);
+        SpigotConfig.readConfig(SpigotWorldConfig.class, this, saveOnLoad);
     }
 
     private void log(String s) {
