From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David <davidliebtkekse@gmail.com>
Date: Thu, 18 Sep 2025 20:28:39 +0200
Subject: [PATCH] World overrides


diff --git a/src/main/java/io/papermc/paper/world/PaperWorldLoader.java b/src/main/java/io/papermc/paper/world/PaperWorldLoader.java
index 1bf392bacaddcccaa5b17872b1646f4eb9876baf..c6cc9b061a392e420f660bbfe44ae419302cb2d7 100644
--- a/src/main/java/io/papermc/paper/world/PaperWorldLoader.java
+++ b/src/main/java/io/papermc/paper/world/PaperWorldLoader.java
@@ -121,11 +121,24 @@ public record PaperWorldLoader(MinecraftServer server, String levelId) {
         final var levelStemRegistry = this.server.registryAccess().lookupOrThrow(Registries.LEVEL_STEM);
         final boolean hasWorldData = this.server.storageSource.hasWorldData();
         final LevelStem overworldStem = requireNonNull(levelStemRegistry.getValue(LevelStem.OVERWORLD), "Overworld stem missing");
-        this.loadInitialWorld(overworldStem, hasWorldData);
+        //ASP start - overwrite vanilla worlds if needed
+        if(!com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadOverworldOverride()) {
+            this.loadInitialWorld(overworldStem, hasWorldData);
+        }
+        //ASP end - overwrite vanilla worlds if needed
         for (final LevelStem stem : levelStemRegistry) {
             if (stem == overworldStem) {
                 continue;
             }
+            //ASP start - overwrite vanilla worlds if needed
+            ResourceKey<LevelStem> levelStemKey = levelStemRegistry.getResourceKey(stem).get();
+            if(levelStemKey == LevelStem.NETHER && com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadNetherOverride()) {
+                continue;
+            }
+            if(levelStemKey == LevelStem.END && com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadEndOverride()) {
+                continue;
+            }
+            //ASP end - overwrite vanilla worlds if needed
             this.loadInitialWorld(stem, hasWorldData);
         }
 
