From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David <davidliebtkekse@gmail.com>
Date: Thu, 18 Sep 2025 20:28:39 +0200
Subject: [PATCH] World overrides


diff --git a/src/main/java/io/papermc/paper/world/PaperWorldLoader.java b/src/main/java/io/papermc/paper/world/PaperWorldLoader.java
index f5e9dd5880681cf4c9b442fbd2e0e9aad1cfb804..6de0f58a8c1a08393276ecaba5b8e7e9640bdbf6 100644
--- a/src/main/java/io/papermc/paper/world/PaperWorldLoader.java
+++ b/src/main/java/io/papermc/paper/world/PaperWorldLoader.java
@@ -3,6 +3,7 @@ package io.papermc.paper.world;
 import com.google.common.io.Files;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Dynamic;
+import net.minecraft.core.Registry;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.nbt.NbtException;
 import net.minecraft.nbt.ReportedNbtException;
@@ -108,7 +109,21 @@ public record PaperWorldLoader(MinecraftServer server, String levelId) {
 
     // Loosely modeled on code in net.minecraft.server.Main
     public void loadInitialWorlds() {
-        for (final LevelStem stem : this.server.registryAccess().lookupOrThrow(Registries.LEVEL_STEM)) {
+        //ASP start - overwrite vanilla worlds if needed
+        Registry<LevelStem> registry = this.server.registryAccess().lookupOrThrow(Registries.LEVEL_STEM);
+        for (final LevelStem stem : registry) {
+            ResourceKey<LevelStem> levelStemKey = registry.getResourceKey(stem).get();
+            if(levelStemKey == LevelStem.NETHER && com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadNetherOverride()) {
+                continue;
+            }
+            if(levelStemKey == LevelStem.END && com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadEndOverride()) {
+                continue;
+            }
+            if(levelStemKey == LevelStem.OVERWORLD && com.infernalsuite.asp.SlimeNMSBridgeImpl.instance().loadOverworldOverride()) {
+                continue;
+            }
+            //ASP end - overwrite vanilla worlds if needed
+
             final WorldLoadingInfo info = this.getWorldInfo(this.levelId, stem);
             this.migrateWorldFolder(info);
             if (!info.enabled()) {
