--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -602,8 +_,35 @@
         this.playerTickingChunks.remove((LevelChunk)chunkHolder.getCurrentChunk());
     }
     // Paper end - chunk tick iteration
-
-    public ServerLevel(
+    public com.infernalsuite.asp.level.SlimeInMemoryWorld slimeInstance; // ASP
+
+    public ServerLevel(
+            final MinecraftServer server,
+            final Executor executor,
+            final LevelStorageSource.LevelStorageAccess levelStorage,
+            final WorldGenSettings worldGenSettings, // CraftBukkit
+            final ResourceKey<Level> dimension,
+            final LevelStem levelStem,
+            final boolean isDebug,
+            final long biomeZoomSeed,
+            final List<CustomSpawner> customSpawners,
+            final boolean tickTime
+            // Paper start - add parameters
+            , ResourceKey<LevelStem> typeKey,
+            org.bukkit.World.Environment env,
+            org.bukkit.generator.ChunkGenerator gen,
+            org.bukkit.generator.BiomeProvider biomeProvider,
+            SavedDataStorage savedDataStorage,
+            io.papermc.paper.world.PaperWorldLoader.LoadedWorldData loadedWorldData
+            // Paper end - add parameters
+    ) {
+        //ASP start
+        this(null, server, executor, levelStorage, worldGenSettings, dimension, levelStem,
+                isDebug, biomeZoomSeed, customSpawners, tickTime, typeKey, env, gen, biomeProvider, savedDataStorage, loadedWorldData);
+    }
+
+    public ServerLevel(
+        com.infernalsuite.asp.level.@Nullable SlimeBootstrap bootstrap,
         final MinecraftServer server,
         final Executor executor,
         final LevelStorageSource.LevelStorageAccess levelStorage,
@@ -623,6 +_,7 @@
         io.papermc.paper.world.PaperWorldLoader.LoadedWorldData loadedWorldData
         // Paper end - add parameters
     ) {
+        //ASP end
         // CraftBukkit start
         final io.papermc.paper.world.saveddata.PaperLevelOverrides levelData = loadedWorldData.levelOverrides();
         savedDataStorage.set(io.papermc.paper.world.saveddata.PaperLevelOverrides.TYPE, levelData);
@@ -644,7 +_,13 @@
         this.server = server;
         this.customSpawners = customSpawners;
         this.serverLevelData = levelData;
+        // ASP START
         ChunkGenerator generator = levelStem.generator();
+        ChunkGenerator result = this.getGenerator(bootstrap);
+        if (result != null) {
+            generator = result;
+        }
+        // ASP END
         // CraftBukkit start
         if (loadedWorldData.pdc() != null) {
             this.getWorld().readBukkitValues(loadedWorldData.pdc().persistentData().toTagCompound());
@@ -732,6 +_,12 @@
         // Paper end - rewrite chunk system
         this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
     }
+
+    // ASP START
+    public @Nullable ChunkGenerator getGenerator(com.infernalsuite.asp.level.@Nullable SlimeBootstrap bootstrap) {
+        return null;
+    }
+    // ASP END
 
     // Paper start
     @Override
