From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Philip Kelley <philip@thoriumcube.org>
Date: Tue, 23 Apr 2024 17:06:36 +0100
Subject: [PATCH] Change shadow plugin


diff --git a/build.gradle.kts b/build.gradle.kts
index 6096e61162231d1168bd072ada18c8a115a6b659..b0c24037ed2ab876e696eccf819e07e67e00ac61 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -4,6 +4,7 @@ import java.time.Instant
 plugins {
     java
     `maven-publish`
+    id("io.github.goooler.shadow")
 }
 
 val log4jPlugins = sourceSets.create("log4jPlugins")
