From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Philip Kelley <philip@thoriumcube.org>
Date: Tue, 23 Apr 2024 17:06:36 +0100
Subject: [PATCH] Change shadow plugin


diff --git a/build.gradle.kts b/build.gradle.kts
index fbdce1cf7cd1c27f94e011e921af48716ae81ea1..088c97e15466e1deba8b2e0bfd7ed7d4757ee81f 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,6 +3,7 @@ import io.papermc.paperweight.util.*
 plugins {
     java
     `maven-publish`
+    id("io.github.goooler.shadow")
 }
 
 val log4jPlugins = sourceSets.create("log4jPlugins")
