From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul19988 <paul@paulcodes.co.uk>
Date: Sat, 12 Aug 2023 22:07:44 +0100
Subject: [PATCH] Build Changes


diff --git a/build.gradle.kts b/build.gradle.kts
index 745bee8b7fbe80cd71e9090b45f41a4baf2a7abc..ddb160884598fae9a2e2e7c942d6e4918238a976 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -12,7 +12,10 @@ configurations.named(log4jPlugins.compileClasspathConfigurationName) {
 val alsoShade: Configuration by configurations.creating
 
 dependencies {
-    implementation(project(":paper-api"))
+    // ASWM start
+    implementation(project(":slimeworldmanager-api"))
+    implementation(project(":core"))
+    // ASWM end
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
