From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kyngs <kyngs@users.noreply.github.com>
Date: Sat, 21 Oct 2023 18:11:15 +0200
Subject: [PATCH] Add support for serializing/deserializing PDC


diff --git a/src/main/java/com/infernalsuite/asp/SlimeNMSBridgeImpl.java b/src/main/java/com/infernalsuite/asp/SlimeNMSBridgeImpl.java
index d1ab18a2a2793807118ae1fc23d21248aa7e09a3..0c1e599aa7d148c40ece1957c89991e1d761162a 100644
--- a/src/main/java/com/infernalsuite/asp/SlimeNMSBridgeImpl.java
+++ b/src/main/java/com/infernalsuite/asp/SlimeNMSBridgeImpl.java
@@ -11,6 +11,7 @@ import com.mojang.serialization.Lifecycle;
 import net.kyori.adventure.util.Services;
 import net.minecraft.SharedConstants;
 import net.minecraft.core.registries.Registries;
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
@@ -178,6 +179,11 @@ public class SlimeNMSBridgeImpl implements SlimeNMSBridge {
         // level.setReady(true);
         level.setSpawnSettings(world.getPropertyMap().getValue(SlimeProperties.ALLOW_MONSTERS), world.getPropertyMap().getValue(SlimeProperties.ALLOW_ANIMALS));
 
+        var nmsExtraData = (CompoundTag) Converter.convertTag(world.getExtraData());
+
+        //Attempt to read PDC
+        if (nmsExtraData.get("BukkitValues") != null) level.getWorld().readBukkitValues(nmsExtraData.get("BukkitValues"));
+
         return level;
     }
 
diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java b/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
index ba337466525971885247c269c23aac7df96d2782..5d325e85a4634c21ab0bec12b493ea9b857d8ff3 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
@@ -2,6 +2,7 @@ package com.infernalsuite.asp.level;
 
 import com.flowpowered.nbt.CompoundTag;
 import com.infernalsuite.asp.ChunkPos;
+import com.infernalsuite.asp.Converter;
 import com.infernalsuite.asp.api.exceptions.WorldAlreadyExistsException;
 import com.infernalsuite.asp.api.loaders.SlimeLoader;
 import com.infernalsuite.asp.serialization.slime.SlimeSerializer;
@@ -235,6 +236,17 @@ public class SlimeInMemoryWorld implements SlimeWorld, SlimeWorldInstance {
             cloned.put(entry.getKey(), clonedChunk);
         }
 
+        // Serialize Bukkit Values (PDC)
+
+        var nmsTag = new net.minecraft.nbt.CompoundTag();
+
+        instance.getWorld().storeBukkitValues(nmsTag);
+
+        // Bukkit stores the relevant tag as a tag with the key "BukkitValues" in the tag we supply to it
+        var flowTag = Converter.convertTag("BukkitValues", nmsTag.getCompound("BukkitValues"));
+
+        world.getExtraData().getValue().put(flowTag);
+
         return new SkeletonSlimeWorld(world.getName(),
                 world.getLoader(),
                 world.isReadOnly(),
