From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul19988 <paul@paulcodes.co.uk>
Date: Sat, 12 Aug 2023 22:07:44 +0100
Subject: [PATCH] Build Changes


diff --git a/build.gradle.kts b/build.gradle.kts
index be5c4212aebed4a932ee2c688b2a31640eca3dc7..d03c9c385d6d969f0de33fa7d8dc547699f64493 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -13,8 +13,13 @@ configurations.named(log4jPlugins.compileClasspathConfigurationName) {
 val alsoShade: Configuration by configurations.creating
 
 dependencies {
-    implementation(project(":paper-api"))
-    implementation(project(":paper-mojangapi"))
+    // ASWM start
+    implementation(project(":slimeworldmanager-api"))
+    implementation(project(":core"))
+    implementation("io.papermc.paper:paper-mojangapi:1.20.4-R0.1-SNAPSHOT") {
+        exclude("io.papermc.paper", "paper-api")
+    }
+    // ASWM end
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
