From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Philip Kelley <philip@thoriumcube.org>
Date: Sat, 4 Nov 2023 20:36:26 +0000
Subject: [PATCH] Handle null data properly


diff --git a/src/main/java/com/infernalsuite/asp/SimpleDataFixerConverter.java b/src/main/java/com/infernalsuite/asp/SimpleDataFixerConverter.java
index 694bd135fd4398ecf02940e638fae1047e217a29..6856a7254685b2d005bcb2a91841f53d8743e38d 100644
--- a/src/main/java/com/infernalsuite/asp/SimpleDataFixerConverter.java
+++ b/src/main/java/com/infernalsuite/asp/SimpleDataFixerConverter.java
@@ -44,9 +44,12 @@ class SimpleDataFixerConverter implements SlimeWorldReader<SlimeWorld> {
                 );
             }
 
+            ChunkPos chunkPos = new ChunkPos(chunk.getX(), chunk.getZ());
+
             SlimeChunkSection[] sections = new SlimeChunkSection[chunk.getSections().length];
             for (int i = 0; i < sections.length; i++) {
                 SlimeChunkSection dataSection = chunk.getSections()[i];
+                if (dataSection == null) continue;
 
                 CompoundBinaryTag blockStateTag = blockStateTag = convertAndBack(dataSection.getBlockStatesTag(), (tag) -> {
                     WalkerUtils.convertList(MCTypeRegistry.BLOCK_STATE, new NBTMapType(tag), "palette", currentVersion, newVersion);
@@ -62,17 +65,17 @@ class SimpleDataFixerConverter implements SlimeWorldReader<SlimeWorld> {
                         dataSection.getBlockLight(),
                         dataSection.getSkyLight()
                 );
-
-                chunks.put(new ChunkPos(chunk.getX(), chunk.getZ()), new SlimeChunkSkeleton(
-                        chunk.getX(),
-                        chunk.getZ(),
-                        sections,
-                        chunk.getHeightMaps(),
-                        blockEntities,
-                        entities
-                ));
             }
 
+            chunks.put(chunkPos, new SlimeChunkSkeleton(
+                    chunk.getX(),
+                    chunk.getX(),
+                    sections,
+                    chunk.getHeightMaps(),
+                    blockEntities,
+                    entities
+            ));
+
         }
 
         return new SkeletonSlimeWorld(
