From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul19988 <paul@paulcodes.co.uk>
Date: Sat, 12 Aug 2023 23:54:39 +0100
Subject: [PATCH] poi data loader


diff --git a/src/main/java/com/infernalsuite/aswm/level/ChunkDataLoadTask.java b/src/main/java/com/infernalsuite/aswm/level/ChunkDataLoadTask.java
index f417056ddcab00a596e65934afe9986acbe14e9d..c32d52c68188dc1eb7feeac364cdc4aded1c4574 100644
--- a/src/main/java/com/infernalsuite/aswm/level/ChunkDataLoadTask.java
+++ b/src/main/java/com/infernalsuite/aswm/level/ChunkDataLoadTask.java
@@ -73,16 +73,6 @@ public final class ChunkDataLoadTask implements CommonLoadTask {
             //                }
 
             LevelChunk chunk = this.world.slimeInstance.promote(chunkX, chunkZ, data);
-            System.out.println("STATUS: " + this.world.slimeInstance.getInstance().chunkSource.getPoiManager());
-
-//            this.world.getPoiManager().getOrLoad(ChunkPos.asLong(this.chunkX, this.chunkZ)).ifPresent(poiSection -> {
-//
-//            });
-//            if (poiChunk == null) {
-//                LOGGER.error("Unable to locate poi for chunk " + this.chunkX + ", " + this.chunkZ);
-//            } else {
-//                poiChunk.load();
-//            }
 
             return new ImposterProtoChunk(chunk, false);
         } catch (final ThreadDeath death) {
