From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Philip Kelley <philip@thoriumcube.org>
Date: Sat, 4 Nov 2023 20:36:26 +0000
Subject: [PATCH] Handle null data properly


diff --git a/src/main/java/com/infernalsuite/aswm/SimpleDataFixerConverter.java b/src/main/java/com/infernalsuite/aswm/SimpleDataFixerConverter.java
index 1affe4c94b490a05184deccc9eb80530f67fd5ea..1a4be97069f01a82deadd26a94e86dbebe0e47a0 100644
--- a/src/main/java/com/infernalsuite/aswm/SimpleDataFixerConverter.java
+++ b/src/main/java/com/infernalsuite/aswm/SimpleDataFixerConverter.java
@@ -45,9 +45,12 @@ class SimpleDataFixerConverter implements SlimeWorldReader<SlimeWorld> {
                 );
             }
 
+            ChunkPos chunkPos = new ChunkPos(chunk.getX(), chunk.getZ());
+
             SlimeChunkSection[] sections = new SlimeChunkSection[chunk.getSections().length];
             for (int i = 0; i < sections.length; i++) {
                 SlimeChunkSection dataSection = chunk.getSections()[i];
+                if (dataSection == null) continue;
 
                 com.flowpowered.nbt.CompoundTag blockStateTag = blockStateTag = convertAndBack(dataSection.getBlockStatesTag(), (tag) -> {
                     WalkerUtils.convertList(MCTypeRegistry.BLOCK_STATE, new NBTMapType(tag), "palette", currentVersion, newVersion);
@@ -63,17 +66,17 @@ class SimpleDataFixerConverter implements SlimeWorldReader<SlimeWorld> {
                         dataSection.getBlockLight(),
                         dataSection.getSkyLight()
                 );
-
-                chunks.put(new ChunkPos(chunk.getX(), chunk.getZ()), new SlimeChunkSkeleton(
-                        chunk.getX(),
-                        chunk.getZ(),
-                        sections,
-                        chunk.getHeightMaps(),
-                        blockEntities,
-                        entities
-                ));
             }
 
+            chunks.put(chunkPos, new SlimeChunkSkeleton(
+                    chunk.getX(),
+                    chunk.getX(),
+                    sections,
+                    chunk.getHeightMaps(),
+                    blockEntities,
+                    entities
+            ));
+
         }
 
         return new SkeletonSlimeWorld(
