From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kyngs <kyngs@users.noreply.github.com>
Date: Sun, 19 May 2024 19:17:07 +0200
Subject: [PATCH] Fix chunks not getting serialized when reloaded.


diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeChunkLevel.java b/src/main/java/com/infernalsuite/asp/level/SlimeChunkLevel.java
index 0d218911fb8be1bc5272b6e291ec60404306c831..a92d17b5c6a7c8427ecbdf56a9c2a0c1b66978e4 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeChunkLevel.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeChunkLevel.java
@@ -24,4 +24,10 @@ public class SlimeChunkLevel extends LevelChunk {
         super.unloadCallback();
         this.inMemoryWorld.unload(this);
     }
+
+    @Override
+    public void loadCallback() {
+        super.loadCallback();
+        this.inMemoryWorld.ensureChunkMarkedAsLoaded(this);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java b/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
index 619075ededa0443cf8f888f04d863bf7b936a4b0..70eb8b95740b46579ad3520ca1d3395a920a8a6a 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
@@ -219,7 +219,7 @@ public class SlimeInMemoryWorld implements SlimeWorld, SlimeWorldInstance {
                     } else {
                         chunk = safeNmsChunkWrapper.getWrapper().getChunk();
                     }
-                } else if  (clonedChunk instanceof NMSSlimeChunk nmsSlimeChunk) {
+                } else if (clonedChunk instanceof NMSSlimeChunk nmsSlimeChunk) {
                     chunk = nmsSlimeChunk.getChunk();
                 }
 
@@ -279,4 +279,10 @@ public class SlimeInMemoryWorld implements SlimeWorld, SlimeWorldInstance {
     public @NotNull PersistentDataContainer getPersistentDataContainer() {
         return this.extraPDC;
     }
+
+    public void ensureChunkMarkedAsLoaded(SlimeChunkLevel chunk) {
+        if (chunkStorage.get(new ChunkPos(chunk.locX, chunk.locZ)) instanceof SlimeChunkSkeleton skeleton) {
+            chunkStorage.put(new ChunkPos(chunk.locX, chunk.locZ), new NMSSlimeChunk(chunk, skeleton));
+        }
+    }
 }
