From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AverageGithub <adorjantoto09@gmail.com>
Date: Sat, 4 May 2024 20:18:16 +0200
Subject: [PATCH] Compile fixes


diff --git a/build.gradle.kts b/build.gradle.kts
index b5b5c099e61f828edbcd18aba9d379a6e45fb8b4..838639a8bbeb59700037c3e101714f68f8b79671 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -29,9 +29,6 @@ dependencies {
     // ASWM start
     implementation(project(":aspaper-api"))
     implementation(project(":core"))
-    implementation("io.papermc.paper:paper-mojangapi:1.20.4-R0.1-SNAPSHOT") {
-        exclude("io.papermc.paper", "paper-api")
-    }
     // ASWM end
     // Paper start
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
diff --git a/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java b/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java
index 419c3c12177bc7b85c6643fb114b4d0ad8f80c46..b6c0ba9325b06e1d76609aa4de564af11f231a74 100644
--- a/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java
+++ b/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java
@@ -46,18 +46,14 @@ public class NMSSlimeChunk implements SlimeChunk {
     static {
         {
             PalettedContainer<BlockState> empty = new PalettedContainer<>(Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES, null);
-            Tag tag = ChunkSerializer.BLOCK_STATE_CODEC.encodeStart(NbtOps.INSTANCE, empty).getOrThrow(false, (error) -> {
-                throw new AssertionError(error);
-            });
+            Tag tag = ChunkSerializer.BLOCK_STATE_CODEC.encodeStart(NbtOps.INSTANCE, empty).getOrThrow();
 
             EMPTY_BLOCK_STATE_PALETTE = (CompoundTag) Converter.convertTag("", tag);
         }
         {
             Registry<Biome> biomes = net.minecraft.server.MinecraftServer.getServer().registryAccess().registryOrThrow(Registries.BIOME);
             PalettedContainer<Holder<Biome>> empty = new PalettedContainer<>(biomes.asHolderIdMap(), biomes.getHolderOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES, null);
-            Tag tag = ChunkSerializer.makeBiomeCodec(biomes).encodeStart(NbtOps.INSTANCE, empty).getOrThrow(false, (error) -> {
-                throw new AssertionError(error);
-            });
+            Tag tag = ChunkSerializer.makeBiomeCodec(biomes).encodeStart(NbtOps.INSTANCE, empty).getOrThrow();
 
             EMPTY_BIOME_PALETTE = (CompoundTag) Converter.convertTag("", tag);
         }
@@ -110,7 +106,7 @@ public class NMSSlimeChunk implements SlimeChunk {
             if (section.hasOnlyAir()) {
                 blockStateTag = EMPTY_BLOCK_STATE_PALETTE;
             } else {
-                Tag data = ChunkSerializer.BLOCK_STATE_CODEC.encodeStart(NbtOps.INSTANCE, section.getStates()).getOrThrow(false, System.err::println); // todo error handling
+                Tag data = ChunkSerializer.BLOCK_STATE_CODEC.encodeStart(NbtOps.INSTANCE, section.getStates()).getOrThrow(); // todo error handling
                 blockStateTag = (CompoundTag) Converter.convertTag("", data);
             }
 
@@ -120,7 +116,7 @@ public class NMSSlimeChunk implements SlimeChunk {
             if (biomes.data.palette().getSize() == 1 && biomes.data.palette().maybeHas((h) -> h.is(Biomes.PLAINS))) {
                 biomeTag = EMPTY_BIOME_PALETTE;
             } else {
-                Tag biomeData = codec.encodeStart(NbtOps.INSTANCE, section.getBiomes()).getOrThrow(false, System.err::println); // todo error handling
+                Tag biomeData = codec.encodeStart(NbtOps.INSTANCE, section.getBiomes()).getOrThrow(); // todo error handling
                 biomeTag = (CompoundTag) Converter.convertTag("", biomeData);
             }
 
@@ -154,7 +150,7 @@ public class NMSSlimeChunk implements SlimeChunk {
         List<net.minecraft.nbt.CompoundTag> tileEntities = new ArrayList<>();
 
         for (BlockEntity entity : chunk.blockEntities.values()) {
-            net.minecraft.nbt.CompoundTag entityNbt = entity.saveWithFullMetadata();
+            net.minecraft.nbt.CompoundTag entityNbt = entity.saveWithFullMetadata(net.minecraft.server.MinecraftServer.getServer().registryAccess());
             tileEntities.add(entityNbt);
         }
 
diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeChunkConverter.java b/src/main/java/com/infernalsuite/asp/level/SlimeChunkConverter.java
index ace4d6272560783fd7903495b5a3c84550b1b4d8..6807767a2067b62558f749da0c25f184082d0769 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeChunkConverter.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeChunkConverter.java
@@ -78,7 +78,7 @@ public class SlimeChunkConverter {
                     DataResult<PalettedContainer<BlockState>> dataresult = ChunkSerializer.BLOCK_STATE_CODEC.parse(NbtOps.INSTANCE, Converter.convertTag(slimeSection.getBlockStatesTag())).promotePartial((s) -> {
                         System.out.println("Recoverable error when parsing section " + x + "," + z + ": " + s); // todo proper logging
                     });
-                    blockPalette = dataresult.getOrThrow(false, System.err::println); // todo proper logging
+                    blockPalette = dataresult.getOrThrow(); // todo proper logging
                 } else {
                     blockPalette = new PalettedContainer<>(Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState(), PalettedContainer.Strategy.SECTION_STATES, null);
                 }
@@ -89,7 +89,7 @@ public class SlimeChunkConverter {
                     DataResult<PalettedContainer<Holder<Biome>>> dataresult = codec.parse(NbtOps.INSTANCE, Converter.convertTag(slimeSection.getBiomeTag())).promotePartial((s) -> {
                         System.out.println("Recoverable error when parsing section " + x + "," + z + ": " + s); // todo proper logging
                     });
-                    biomePalette = dataresult.getOrThrow(false, System.err::println); // todo proper logging
+                    biomePalette = dataresult.getOrThrow(); // todo proper logging
                 } else {
                     biomePalette = new PalettedContainer<>(biomeRegistry.asHolderIdMap(), biomeRegistry.getHolderOrThrow(Biomes.PLAINS), PalettedContainer.Strategy.SECTION_BIOMES, null);
                 }
@@ -137,7 +137,7 @@ public class SlimeChunkConverter {
                 if (type.isPresent()) {
                     BlockPos blockPosition = new BlockPos(tag.getIntValue("x").get(), tag.getIntValue("y").get(), tag.getIntValue("z").get());
                     BlockState blockData = nmsChunk.getBlockState(blockPosition);
-                    BlockEntity entity = BlockEntity.loadStatic(blockPosition, blockData, (net.minecraft.nbt.CompoundTag) Converter.convertTag(tag));
+                    BlockEntity entity = BlockEntity.loadStatic(blockPosition, blockData, (net.minecraft.nbt.CompoundTag) Converter.convertTag(tag), net.minecraft.server.MinecraftServer.getServer().registryAccess());
 
                     if (entity != null) {
                         nmsChunk.setBlockEntity(entity);
diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeLevelGenerator.java b/src/main/java/com/infernalsuite/asp/level/SlimeLevelGenerator.java
index b74ccd2b6b526f7b7955ce601a37185d3b6a2f59..18053d099145bd9351902ff062139c731eafc5d3 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeLevelGenerator.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeLevelGenerator.java
@@ -1,6 +1,7 @@
 package com.infernalsuite.asp.level;
 
 import com.mojang.serialization.Codec;
+import com.mojang.serialization.MapCodec;
 import net.minecraft.core.Holder;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeSource;
@@ -21,7 +22,7 @@ public class SlimeLevelGenerator extends FlatLevelSource {
     private static BiomeSource getSource(Holder<Biome> biome) {
         return new BiomeSource() {
             @Override
-            protected Codec<? extends BiomeSource> codec() {
+            protected MapCodec<? extends BiomeSource> codec() {
                 return null;
             }
 
diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeLevelInstance.java b/src/main/java/com/infernalsuite/asp/level/SlimeLevelInstance.java
index eddae8fc510d6e263df55a725148d96f677bcb0b..a8c103d435ac9473b1c5dd404e5c69ebf4353490 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeLevelInstance.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeLevelInstance.java
@@ -103,8 +103,6 @@ public class SlimeLevelInstance extends ServerLevel {
         super.setSpawnSettings(propertyMap.getValue(SlimeProperties.ALLOW_MONSTERS), propertyMap.getValue(SlimeProperties.ALLOW_ANIMALS));
 
         this.pvpMode = propertyMap.getValue(SlimeProperties.PVP);
-
-        this.keepSpawnInMemory = false;
     }
 
     @Override
@@ -123,7 +121,7 @@ public class SlimeLevelInstance extends ServerLevel {
 
                 //this.getChunkSource().save(forceSave);
                 this.serverLevelData.setWorldBorder(this.getWorldBorder().createSettings());
-                this.serverLevelData.setCustomBossEvents(MinecraftServer.getServer().getCustomBossEvents().save());
+                this.serverLevelData.setCustomBossEvents(MinecraftServer.getServer().getCustomBossEvents().save(MinecraftServer.getServer().registryAccess()));
 
                 // Update level data
                 net.minecraft.nbt.CompoundTag compound = new net.minecraft.nbt.CompoundTag();
