From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: evlad <emmanuelvlad@gmail.com>
Date: Thu, 19 Sep 2024 20:47:32 +0300
Subject: [PATCH] fix chunk pdc getting wiped on chunk unload


diff --git a/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java b/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java
index 97e96f51a6369a328a46086ebb250c9c163e10cf..8053d224ca673c3abe60151b3b322e3e4c9c3fcf 100644
--- a/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java
+++ b/src/main/java/com/infernalsuite/asp/level/NMSSlimeChunk.java
@@ -68,6 +68,7 @@ public class NMSSlimeChunk implements SlimeChunk {
     public NMSSlimeChunk(LevelChunk chunk, SlimeChunk reference) {
         this.chunk = chunk;
         this.extra = reference == null ? CompoundBinaryTag.empty() : reference.getExtraData();
+        this.extra.put("ChunkBukkitValues", Converter.convertTag(chunk.persistentDataContainer.toTagCompound()));
         this.upgradeData = reference == null ? null : reference.getUpgradeData();
     }
 
