From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Philip Kelley <philip@thoriumcube.org>
Date: Fri, 20 Dec 2024 11:31:09 +0000
Subject: [PATCH] Remove leftover FlowNBT usage


diff --git a/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java b/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
index 166f8c17df110d4ed13dd31a9de06c3d7b5c070b..a5f35f94d82928f120f02909086f57d4ceaa52c2 100644
--- a/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
+++ b/src/main/java/com/infernalsuite/asp/level/SlimeInMemoryWorld.java
@@ -113,8 +113,8 @@ public class SlimeInMemoryWorld implements SlimeWorld, SlimeWorldInstance {
             this.chunkStorage.remove(Util.chunkPosition(x, z));
             return;
         }
-        Tag<?> pdcTag = Converter.convertTag("ChunkBukkitValues", providedChunk.persistentDataContainer.toTagCompound());
-        chunk.getExtraData().getValue().put(pdcTag);
+        CompoundBinaryTag pdcTag = Converter.convertTag(providedChunk.persistentDataContainer.toTagCompound());
+        chunk.getExtraData().put("ChunkBukkitValues", pdcTag);
 
         this.chunkStorage.put(Util.chunkPosition(x, z),
                 new SlimeChunkSkeleton(chunk.getX(), chunk.getZ(), chunk.getSections(),
